internal/webhook/external/handlers/refresh_warehouse.go

```diff
@@ -0,0 +1,104 @@
+package handlers
+
+import (
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/api"
+	xhttp "github.com/akuity/kargo/internal/http"
+	"github.com/akuity/kargo/internal/indexer"
+	"github.com/akuity/kargo/internal/logging"
+	"github.com/akuity/kargo/internal/webhook/external/providers"
+)
+
+// NewRefreshWarehouseWebhook handles push events for the designated provider.
+// After the provider has been resolved and the request has been authenticated,
+// the kubeclient is queried for all warehouses that contain a subscription
+// to the repo in question. Those warehouses are then patched with a special
+// annotation that signals down stream logic to refresh the warehouse.
+func NewRefreshWarehouseWebhook(p providers.Provider, c client.Client) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
+		repo, err := p.GetRepository(r)
+		if err != nil {
+			code := xhttp.CodeFrom(err)
+			xhttp.WriteErrorf(w,
+				code,
+				"failed to get repository: %w",
+				err,
+			)
```

Comment:
I like how you're getting the response code from the error. 👍 

```suggestion
			xhttp.WriteErrorf(
				w,
				xhttp.CodeFrom(err),
				"failed to get repository: %w",
				err,
			)
```

Reply:
Actually... thinking about it more, all these `xhttp.WriteX()` seem as if they could be mapping errs to status codes internally instead of requiring the caller to find the code first and provide it as a separate arg.